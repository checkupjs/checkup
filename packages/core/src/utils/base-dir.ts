export const BASE_DIR = process.cwd();
