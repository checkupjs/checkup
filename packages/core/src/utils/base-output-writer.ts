import * as chalk from 'chalk';
import { startCase } from 'lodash';
import CliTable3 = require('cli-table3');
import CheckupError from '../errors/checkup-error';

type Segment = { title: string; count: number; color?: chalk.Chalk };

export default abstract class BaseOutputWriter {
  abstract write(content: string): void;

  blankLine() {
    this.write('\n');
  }

  clearScreen() {
    this.write('\u001B[2J');
  }

  clearLine() {
    this.write('\u001B[0f');
  }

  categoryHeader(header: string) {
    this.write(chalk.dim('=== ') + chalk.bold(header) + '\n');
    this.blankLine();
  }

  sectionHeader(header: string) {
    this.write(this.emphasize(`${chalk.underline(startCase(header))}\n`));
    this.blankLine();
  }

  section(header: string | undefined, contents: () => void) {
    if (header) {
      this.sectionHeader(header);
      contents();
      this.blankLine();
    }
  }

  subHeader(header: string) {
    this.write(`${chalk.underline(startCase(header))}\n`);
    this.blankLine();
  }

  log(value: string) {
    this.write(value);
    this.blankLine();
  }

  dimmed(format: string) {
    this.log(chalk.dim(format));
  }

  emphasize(format: string) {
    return chalk.bold(format);
  }

  bar(title: string, complete: number, total: number, unit: string = '', maximum: number = 50) {
    const barTick = '■';
    const barSegment = Math.ceil(total / maximum);
    const completedSegments = Math.ceil(complete / barSegment);
    const incompleteSegments = maximum - completedSegments;
    const bar = `${this.randomColor()(barTick.repeat(completedSegments))}${chalk.grey(
      barTick.repeat(Math.max(0, incompleteSegments))
    )}`;

    this.log(title);
    this.log(`${bar} ${complete.toLocaleString()}${unit}`);
    this.blankLine();
  }

  table(rows: (string | number)[][], headers: string[]) {
    const table = new CliTable3({
      head: headers.map((header) => this.emphasize(chalk.white(header))),
      chars: {
        top: '',
        'top-mid': '',
        'top-left': '',
        'top-right': '',
        bottom: '',
        'bottom-mid': '',
        'bottom-left': '',
        'bottom-right': '',
        left: '',
        'left-mid': '',
        mid: '',
        'mid-mid': '',
        right: '',
        'right-mid': '',
        middle: ' ',
      },
      style: { 'padding-left': 0, 'padding-right': 0 },
    });

    rows.forEach((row) => {
      table.push(row);
    });

    this.log(table.toString());
  }

  sectionedBar(segments: Segment[], total: number, unit: string = '', width: number = 50) {
    const barTick = '■';
    const colors = this.colors();

    segments.forEach((segment, i) => {
      segment.color = colors[i + 3];
    });

    const { completedSegments, incompleteSegments } = calculateSectionBar(segments, total, width);

    const bar = `${completedSegments
      .map((segment) => segment.color!(barTick.repeat(segment.completed)))
      .join('')}${chalk.grey(barTick.repeat(Math.max(0, incompleteSegments)))}`;

    if (incompleteSegments > 0) {
      completedSegments.push({
        title: 'unknown',
        completed: incompleteSegments,
        count: incompleteSegments,
        color: chalk.grey,
      });
    }

    this.log(`${bar} ${total.toLocaleString()} ${unit}`);
    completedSegments.map((segment) =>
      this.log(`${segment.color!(barTick)} ${segment.title} (${segment.count.toLocaleString()})`)
    );
  }

  valuesList(
    values: { title: string; count: number }[],
    unit: string = '',
    color: chalk.Chalk = this.randomColor()
  ) {
    values.forEach((value) => {
      this.value(value, unit, color);
    });
  }

  styledJSON(object: any) {
    const json = JSON.stringify(object, undefined, 2);
    this.log(json);
  }

  value(
    value: { title: string; count: number },
    unit: string = '',
    color: chalk.Chalk = this.randomColor()
  ) {
    this.log(
      `${color('■')} ${value.title} (${value.count.toLocaleString()}${unit ? ' ' + unit : ''})`
    );
  }

  randomColor() {
    const colors = this.colors();
    return colors[Math.floor(Math.random() * colors.length)];
  }

  colors(range: number = 70) {
    const ANSI_CODE_START = 33;

    return Array.from({ length: range }).map((_, i) => chalk.ansi256(i + ANSI_CODE_START));
  }

  getColor(color: string | chalk.Chalk) {
    if (typeof color === 'string') {
      return chalk.keyword(color);
    }
    return color;
  }

  error(error: string | Error | CheckupError) {
    if (typeof error === 'string') {
      error = new Error(error);
    }

    console.error(
      error instanceof CheckupError ? error.render() : `\n${chalk.red('Error')} ${error.message}`
    );
  }
}

export function calculateSectionBar(
  segments: { title: string; count: number; color?: chalk.Chalk }[],
  total: number,
  width: number
) {
  // We want to normalize the amount provided by the proper ratios.
  // If the total is less than the width we want to multiply the ratio.
  // If the total is greater than the width we want to divide the amount vs the ratio
  const normalizeSegment = function (amount: number) {
    return Math.ceil(total < width ? amount * (width / total) : amount / (total / width));
  };

  const completedSegments = segments
    .map((segment) => {
      return Object.assign({}, segment, {
        completed: normalizeSegment(segment.count),
      });
    })
    .sort((a: Segment, b: Segment) => {
      if (a.count === b.count) {
        return a.title > b.title ? 1 : -1;
      }
      return b.count - a.count;
    });

  // remove  the segments counts from the total number and then normalize it by the barSegment ratio
  const incompleteSegments: number = normalizeSegment(
    total - completedSegments.reduce((prev, curr) => prev + curr.count, 0)
  );

  // if incompleteSegments is a negative number resulting from overflow we return 0
  return {
    completedSegments,
    incompleteSegments: incompleteSegments > 0 ? incompleteSegments : 0,
  };
}
