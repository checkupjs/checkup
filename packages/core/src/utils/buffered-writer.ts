import stripAnsi from 'strip-ansi';
import BaseOutputWriter from './base-output-writer.js';

export default class BufferedWriter extends BaseOutputWriter {
  buffer: string = '';

  get escapedBuffer() {
    return stripAnsi(this.buffer);
  }

  write(content: string): void {
    this.buffer += content;
  }
}
