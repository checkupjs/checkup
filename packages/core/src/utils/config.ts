import { existsSync, readJsonSync, writeJsonSync } from 'fs-extra';
import { join, resolve } from 'path';

import { CheckupConfig } from '../types/config';

const debug = require('debug')('checkup:config');

export const DEFAULT_CONFIG: CheckupConfig = {
  plugins: [],
  tasks: {},
};

export function getConfigPath(path: string = '') {
  path = path || process.cwd();
  return join(path, '.checkuprc');
}

export function readConfig(configPath: string) {
  let config: CheckupConfig;

  debug(`Reading config from ${configPath}`);

  try {
    config = readJsonSync(resolve(configPath || getConfigPath()));
  } catch (error) {
    throw new Error(`Could not locate a Checkup config in ${configPath}`);
  }

  debug(`Found config %o`, config);

  return config;
}

export function mergeConfig(config: Partial<CheckupConfig>) {
  return { ...DEFAULT_CONFIG, ...config };
}

export function writeConfig(path: string = getConfigPath(), config: Partial<CheckupConfig> = {}) {
  if (existsSync(path)) {
    throw new Error(`There is already an existing Checkup config in ${path}`);
  }

  try {
    writeJsonSync(path, mergeConfig(config), { spaces: 2 });
  } catch (error) {
    throw new Error(`Cannot write Checkup config to ${path}. ${error.message}`);
  }

  return path;
}
