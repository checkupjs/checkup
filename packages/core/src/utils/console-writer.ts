import BaseOutputWriter from './base-output-writer.js';
export default class ConsoleWriter extends BaseOutputWriter {
  write(content: string): void {
    process.stdout.write(content);
  }
}
