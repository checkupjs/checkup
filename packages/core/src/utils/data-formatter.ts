export function wrapEntries(
  data: Record<string, string>,
  schema: { keyName: string; valueName: string }
): object[] {
  let result = [];

  for (let key in data) {
    result.push({ [schema.keyName]: key, [schema.valueName]: data[key] });
  }

  return result;
}
