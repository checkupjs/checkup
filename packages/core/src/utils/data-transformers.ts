import * as startCase from 'startcase';

import { TaskItemData } from '../types/tasks';
/**
 * @param data
 * @param schema
 */
export function toPairs(
  data: Record<string, string>,
  schema: { keyName: string; valueName: string }
): object[] {
  let result = [];

  for (let key in data) {
    result.push({ [schema.keyName]: key, [schema.valueName]: data[key] });
  }

  return result;
}

export function toTaskData(results: [string, string[] | Record<string, string>][]): TaskItemData[] {
  return results.map(([type, data]) => {
    return toTaskItemData(type, data);
  });
}

export function toTaskItemData(
  type: string,
  data: string[] | Record<string, string>
): TaskItemData {
  return {
    displayName: startCase(type),
    type,
    data,
    total: Array.isArray(data) ? data.length : Object.keys(data).length,
  };
}
