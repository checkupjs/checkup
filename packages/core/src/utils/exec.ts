const util = require('util');
const execAsPromise = util.promisify(require('child_process').exec);

/**
 * @param {string} cmd - The command to run
 * @param {any} options - Options passed to the command
 * @param {string|number} defaultValue - Default value returned if the command returns no value
 * @param {Function} toType - A function used to convert a result to a specific type
 * @returns {string|number} - The result of the command
 */
export async function exec(
  cmd: string,
  options: any,
  defaultValue: string | number,
  toType: Function = String
) {
  const { stdout } = await execAsPromise(cmd, options);

  return toType(stdout.trim()) || defaultValue;
}
