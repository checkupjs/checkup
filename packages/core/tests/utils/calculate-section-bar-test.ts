import { calculateSectionBar } from '../../src/utils/base-output-writer';

describe('calculate-sectioned-bar', () => {
  describe('calculateSectionBar', function () {
    it('should calculateSectionBar segments correctly', function () {
      expect(
        calculateSectionBar(
          [
            {
              title: 'foo',
              count: 5,
            },
            {
              title: 'bar',
              count: 495,
            },
          ],
          500,
          50
        )
      ).toMatchSnapshot();
      expect(
        calculateSectionBar(
          [
            {
              title: 'foo',
              count: 5,
            },

            {
              title: 'bar',
              count: 10,
            },
          ],

          20,
          50
        )
      ).toMatchInlineSnapshot(
        {},
        `
        {
          "completedSegments": [
            {
              "completed": 25,
              "count": 10,
              "title": "bar",
            },
            {
              "completed": 13,
              "count": 5,
              "title": "foo",
            },
          ],
          "incompleteSegments": 13,
        }
      `
      );
      expect(
        calculateSectionBar(
          [
            {
              title: 'foo',
              count: 10,
            },
            {
              title: 'bar',
              count: 10,
            },
          ],
          20,
          50
        )
      ).toMatchSnapshot();
      expect(
        calculateSectionBar(
          [
            {
              title: 'foo',
              count: 10,
            },
            {
              title: 'moo',
              count: 10,
            },
            {
              title: 'bar',
              count: 10,
            },
          ],
          30,
          50
        )
      ).toMatchSnapshot();
      expect(
        calculateSectionBar(
          [
            {
              title: 'foo',
              count: 50,
            },
            {
              title: 'foo',
              count: 50,
            },
            {
              title: 'bar',
              count: 50,
            },
          ],
          150,
          50
        )
      ).toMatchSnapshot();
      expect(
        calculateSectionBar(
          [
            {
              title: 'foo',
              count: 50,
            },
            {
              title: 'foo',
              count: 50,
            },
            {
              title: 'bar',
              count: 20,
            },
          ],
          150,
          50
        )
      ).toMatchSnapshot();
    });
  });
});
