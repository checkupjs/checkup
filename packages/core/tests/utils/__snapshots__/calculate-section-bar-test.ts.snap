// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`calculate-sectioned-bar > calculateSectionBar > should calculateSectionBar segments correctly 1`] = `
{
  "completedSegments": [
    {
      "completed": 50,
      "count": 495,
      "title": "bar",
    },
    {
      "completed": 1,
      "count": 5,
      "title": "foo",
    },
  ],
  "incompleteSegments": 0,
}
`;

exports[`calculate-sectioned-bar > calculateSectionBar > should calculateSectionBar segments correctly 3`] = `
{
  "completedSegments": [
    {
      "completed": 25,
      "count": 10,
      "title": "bar",
    },
    {
      "completed": 25,
      "count": 10,
      "title": "foo",
    },
  ],
  "incompleteSegments": 0,
}
`;

exports[`calculate-sectioned-bar > calculateSectionBar > should calculateSectionBar segments correctly 4`] = `
{
  "completedSegments": [
    {
      "completed": 17,
      "count": 10,
      "title": "bar",
    },
    {
      "completed": 17,
      "count": 10,
      "title": "foo",
    },
    {
      "completed": 17,
      "count": 10,
      "title": "moo",
    },
  ],
  "incompleteSegments": 0,
}
`;

exports[`calculate-sectioned-bar > calculateSectionBar > should calculateSectionBar segments correctly 5`] = `
{
  "completedSegments": [
    {
      "completed": 17,
      "count": 50,
      "title": "bar",
    },
    {
      "completed": 17,
      "count": 50,
      "title": "foo",
    },
    {
      "completed": 17,
      "count": 50,
      "title": "foo",
    },
  ],
  "incompleteSegments": 0,
}
`;

exports[`calculate-sectioned-bar > calculateSectionBar > should calculateSectionBar segments correctly 6`] = `
{
  "completedSegments": [
    {
      "completed": 17,
      "count": 50,
      "title": "foo",
    },
    {
      "completed": 17,
      "count": 50,
      "title": "foo",
    },
    {
      "completed": 7,
      "count": 20,
      "title": "bar",
    },
  ],
  "incompleteSegments": 10,
}
`;
